Nightly Cryptol regression: well-typed program that shadows variable name no longer typechecks

The expression `\(x : [16][8]) -> map (\x -> x) x == x` should be well-typed in Cryptol, even though the inner `(\x -> x)` expression's `x` variable shadows the outermost `x` binding. In Cryptol 3.5.0, this expression typechecks:

```
Cryptol> :prove \(x : [16][8]) -> map (\x -> x) x == x
[warning] at <interactive>:1:32--1:33
    This binding for `x` shadows the existing binding at
    <interactive>:1:10--1:11
Q.E.D.
(Total Elapsed Time: 0.009s, using "Z3")
```

With the `master` branch of Cryptol (at commit aa70cab02ca18c54fdb77b30e01a94238353b4ab), however, this no longer typechecks:

```
Cryptol> :prove \(x : [16][8]) -> map (\x -> x) x == x
[warning] at <interactive>:1:32--1:33
    Unused name: x
[warning] at <interactive>:1:32--1:33
    This binding for `x` shadows the existing binding at
    <interactive>:1:10--1:11

[error] at <interactive>:1:45--1:46:
  Type mismatch:
    Expected type: 16
    Inferred type: 8
    Context: [_] [ERROR] _
    When checking type of function argument
[error] at <interactive>:1:45--1:46:
  Type mismatch:
    Expected type: [8]
    Inferred type: Bit
    Context: [_] [_] ERROR
    When checking type of function argument
```

Note that if I change the inner `x` to some other name (e.g., `\(x : [16][8]) -> map (\y -> y) x == x`), then the expression typechecks again.

--- a/src/Cryptol/ModuleSystem/Base.hs
+++ b/src/Cryptol/ModuleSystem/Base.hs
@@ -681,7 +681,7 @@ checkModule isrc m = do
 
   -- remove pattern bindings
   npm <- noPat m
-  dump PassNoPat (pp m) 
+  dump PassNoPat (pp npm) 
 
   -- run expandPropGuards
   epgm <- expandPropGuards npm

--- a/src/Cryptol/ModuleSystem/Renamer.hs
+++ b/src/Cryptol/ModuleSystem/Renamer.hs
@@ -1314,7 +1314,7 @@ instance Rename Expr where
 instance Rename FunDesc where
   rename (FunDesc nm offset fromP) =
     do
-      nm' <- traverse resolveCurBind nm
+      nm' <- traverse (resolveCurBind fromP) nm
       pure (FunDesc nm' offset fromP)
 
 --------------------------------------------------------------------------------

--- a/src/Cryptol/ModuleSystem/Renamer.hs
+++ b/src/Cryptol/ModuleSystem/Renamer.hs
@@ -1197,7 +1197,6 @@ maybeLoc mb e =
 -- | Rename a binding.
 instance Rename Bind where
   rename b =
-    doBind
     do
       n'    <- rnLocated (resolveNameDef NSValue) (bName b)
       let checkSig k =
@@ -1207,7 +1206,7 @@ instance Rename Bind where
       checkSig \mbSig ->
         renameBindParams (bParams b) \ps ->
         do
-          e' <- rnLocated rename (bDef b)
+          e' <- setCurBind (bName b) (thing n') (rnLocated rename (bDef b))
           pure b {
             bName      = n',
             bParams    = ps,
@@ -1315,13 +1314,8 @@ instance Rename Expr where
 instance Rename FunDesc where
   rename (FunDesc nm offset fromP) =
     do
-      env <- getLastBindDefs
-      nm' <- forM nm \x ->
-        case lookupNS NSValue x env of
-          Just a -> pure (anyOne a)
-          Nothing -> panic "Rename FunDesc" ["Missing"]
+      nm' <- traverse resolveCurBind nm
       pure (FunDesc nm' offset fromP)
-
 
 --------------------------------------------------------------------------------
 -- Records

--- a/src/Cryptol/ModuleSystem/Renamer/Monad.hs
+++ b/src/Cryptol/ModuleSystem/Renamer/Monad.hs
@@ -164,12 +164,13 @@ data RO = RO {
   curBind :: Maybe (Located PName, Name),
   -- ^ The current binding we are working on.  During NoPat we do
   -- a transformation like this:
-  -- f x y = e   ~>   f = \(f) x -> \(f) y -> e
+  -- f x y = e   ~>   f = \/*f*/ x -> \/*f*/ y -> e
   -- We use `curBind` to resolve the `f` in the lambdas.  Specifically,
   -- we want to avoid getting confused in a situation like this:
-  -- f f y =     ~>   f = \(f) f -> \(f) y -> e  
+  -- f f y =     ~>   f = \/*f*/ f -> \/*f*/ y -> e  
   -- If we are not careful the (f) in the second lambda could refer to the
-  -- argument instead of the original binding.  
+  -- argument instead of the original binding.
+  -- The /*f*/ is the 'FunDesc' in the 'EFun'
 
   localBindEnv :: NamingEnv,
   -- ^ Local names that are in scope, for resolving definition names

--- a/src/Cryptol/ModuleSystem/Renamer/Monad.hs
+++ b/src/Cryptol/ModuleSystem/Renamer/Monad.hs
@@ -586,13 +586,17 @@ setCurBind :: Located PName -> Name -> RenameM a -> RenameM a
 setCurBind p n (R m) = R (mapReader upd m)
   where upd r = r { curBind = Just (p,n) }
 
-resolveCurBind :: PName -> RenameM Name
-resolveCurBind p =
+resolveCurBind :: Bool -> PName -> RenameM Name
+resolveCurBind fromP p =
   do
     mb <- R (curBind <$> ask)
     case mb of
       Just (p',n)
         | thing p' == p -> pure n
+        | fromP,
+          let i = identText (getIdent (thing p')),
+          let j = identText (getIdent p),
+          j `Text.isPrefixOf` i -> pure n
         | otherwise ->
           panic "resolveCurBind"
                 [ "Unexpected current binding"

--- a/src/Cryptol/ModuleSystem/Renamer/Monad.hs
+++ b/src/Cryptol/ModuleSystem/Renamer/Monad.hs
@@ -34,8 +34,9 @@ module Cryptol.ModuleSystem.Renamer.Monad
   , inSubmodule
   , inLocalScope
   , inLocalBindScope
-  , doBind
-  , getLastBindDefs
+
+  , setCurBind
+  , resolveCurBind
 
     -- * Name generation
   , doDefGroup
@@ -70,6 +71,7 @@ import Data.Text qualified as Text
 
 import Cryptol.Utils.Panic
 import Cryptol.Utils.Ident
+import Cryptol.Utils.PP(pp,(<+>))
 import Cryptol.Parser.Name
 import Cryptol.Parser.Position
 import Cryptol.Parser.AST
@@ -111,7 +113,7 @@ runRenamer info (R m) = (res, reverse (renWarnings rwFin))
         es -> Left (reverse es)
 
     ro0 = RO {
-      lastBindsEnv = mempty,
+      curBind = Nothing,
       localsEnv = mempty,
       localBindEnv = mempty,
       outEnv = renEnv info,
@@ -159,21 +161,21 @@ data RO = RO {
   -- We keep this so then if one of the modules is imported, we can
   -- collect its definitions in `externalDeps` to give the typechecker.
 
-  lastBindsEnv :: NamingEnv,
-  -- ^ This is the last environment that introduced function bindings.
-  -- It is basically the value of `getCurBinds` at the point we start
-  -- resolving a binding.  We need this, because during `NoPat`, we desugar
-  -- bindings with arguments, into lambdas, but those lambdas are annotated
-  -- with the name of the function that gave rise to them.  We use this
-  -- environment to avoid resolving this name, thus avoiding confusion with
-  -- the parameters of the function.
+  curBind :: Maybe (Located PName, Name),
+  -- ^ The current binding we are working on.  During NoPat we do
+  -- a transformation like this:
+  -- f x y = e   ~>   f = \(f) x -> \(f) y -> e
+  -- We use `curBind` to resolve the `f` in the lambdas.  Specifically,
+  -- we want to avoid getting confused in a situation like this:
+  -- f f y =     ~>   f = \(f) f -> \(f) y -> e  
+  -- If we are not careful the (f) in the second lambda could refer to the
+  -- argument instead of the original binding.  
 
   localBindEnv :: NamingEnv,
-  -- ^ Used to resolve the names of definitions in the current scope.
+  -- ^ Local names that are in scope, for resolving definition names
 
   localsEnv :: NamingEnv,
-  -- ^ Variables local to a function. These are variables from outside the
-  -- current binding scope.
+  -- ^ Local names that are in scope, for resolving name uses
 
   outEnv :: NamingEnv,
   -- ^ Things in an enclosing scope (for nested modules).  This is used
@@ -520,8 +522,7 @@ getCurBinds = R
     pure (localBindEnv ro `shadowing` defEnv rw)
 
 
--- | Compute the current scope, including local definitions, outer environment
--- and imports.
+-- | Compute the current scope, for resolving name uses.
 getCurScope :: RenameM NamingEnv
 getCurScope = R
   do
@@ -529,7 +530,6 @@ getCurScope = R
     rw <- get
     pure $
       localsEnv ro `shadowing`
-      localBindEnv ro `shadowing`
       defEnv    rw `shadowing`
       modParamEnv (modParams rw) `shadowing`
       impEnv    rw `shadowing`
@@ -582,16 +582,25 @@ addImported rng env =
     forM_ (findShadowing env outDs) \(_,_,xs) ->
       addWarning (SymbolShadowed (ImportShadower rng) xs)
 
--- | Capture the `getCurBinds` at this point.  See the comment
--- on `lastBindsEnv` for more info.
-doBind :: RenameM a -> RenameM a
-doBind (R m) =
-  do
-    env <- getCurBinds
-    R (mapReader (\ro -> ro { lastBindsEnv = env }) m)
+setCurBind :: Located PName -> Name -> RenameM a -> RenameM a
+setCurBind p n (R m) = R (mapReader upd m)
+  where upd r = r { curBind = Just (p,n) }
 
-getLastBindDefs :: RenameM NamingEnv
-getLastBindDefs = R (lastBindsEnv <$> ask)
+resolveCurBind :: PName -> RenameM Name
+resolveCurBind p =
+  do
+    mb <- R (curBind <$> ask)
+    case mb of
+      Just (p',n)
+        | thing p' == p -> pure n
+        | otherwise ->
+          panic "resolveCurBind"
+                [ "Unexpected current binding"
+                , "Expected: " ++ show (pp (thing p') <+> pp (srcRange p'))
+                , "Actual: " ++ show (pp p)
+                ]
+      Nothing ->
+          panic "resolveCurBind" ["No current binding", "Actual: " ++ show (pp p) ]
 
 -- | Set the names of bindings for the duration of a computation.
 inLocalBindScope :: Bool -> NamingEnv -> RenameM a -> RenameM a
@@ -607,7 +616,7 @@ inLocalBindScope checkUsed env (R m) =
   where
   upd ro = ro {
     localBindEnv = env,
-    localsEnv = localBindEnv ro `shadowing` localsEnv ro
+    localsEnv = env `shadowing` localsEnv ro
   }
 
 -- | Do something that will only modify the local scope, and restore

--- a/src/Cryptol/Parser/AST.hs
+++ b/src/Cryptol/Parser/AST.hs
@@ -1431,8 +1431,11 @@ instance (Show name, PPName name) => PP (Expr name) where
       ESel    e l   -> ppPrec 4 e <.> text "." <.> pp l
 
       -- low prec
-      EFun _ xs e   -> wrap n 0 ((text "\\" <.> hsep (map (ppPrec 3) xs)) <+>
+      EFun d xs e   -> wrap n 0 ((text "\\" <.> des <.> hsep (map (ppPrec 3) xs)) <+>
                                  text "->" </> pp e)
+        where
+        des = maybe mempty ppNm (funDescrName d)
+        ppNm x = "/*" <.> pp x <.> "*/"
 
       EIf e1 e2 e3  -> wrap n 0 $ sep [ text "if"   <+> pp e1
                                       , text "then" <+> pp e2
@@ -1476,6 +1479,9 @@ instance (Show name, PPName name) => PP (Expr name) where
    isInfix _ = Nothing
    prefixText PrefixNeg        = "-"
    prefixText PrefixComplement = "~"
+
+instance (Show name, PP name) => PP (FunDesc name) where
+  ppPrec _ fd = maybe mempty pp (funDescrName fd)
 
 instance (Show name, PPName name) => PP (CaseAlt name) where
   ppPrec _ (CaseAlt p e) = vcat [ pp p <+> "->", nest 2 (pp e) ]

--- a/tests/issues/issue2085.cry
+++ b/tests/issues/issue2085.cry
@@ -2,3 +2,8 @@
 bad = \(x : [16][8]) -> map (\x -> x) x == x
 
 example example y = example + y
+
+fun : {n} (fin n) => [n] -> [n]
+fun x
+  | n == 0 => 0
+  | n > 0 => 1

--- a/tests/issues/issue2085.cry
+++ b/tests/issues/issue2085.cry
@@ -1,0 +1,4 @@
+
+bad = \(x : [16][8]) -> map (\x -> x) x == x
+
+example example y = example + y

--- a/tests/issues/issue2085.icry
+++ b/tests/issues/issue2085.icry
@@ -1,0 +1,1 @@
+:load issue2085.cry

--- a/tests/issues/issue2085.icry.stdout
+++ b/tests/issues/issue2085.icry.stdout
@@ -1,0 +1,9 @@
+Loading module Cryptol
+Loading module Cryptol
+Loading module Main
+[warning] at issue2085.cry:2:31--2:32
+    This binding for `x` shadows the existing binding at
+    issue2085.cry:2:9--2:10
+[warning] at issue2085.cry:4:9--4:16
+    This binding for `example` shadows the existing binding at
+    issue2085.cry:4:1--4:8

--- a/tests/issues/issue567.icry.stdout
+++ b/tests/issues/issue567.icry.stdout
@@ -1,24 +1,24 @@
 Loading module Cryptol
 [warning] at issue567.icry:1:7--1:8
-    Unused name: x
-[warning] at issue567.icry:1:7--1:8
     This binding for `x` shadows the existing binding at
     issue567.icry:1:5--1:6
-(\x x -> x : [8]) : {a} [8] -> a -> [8]
-[warning] at issue567.icry:2:11--2:12
+[warning] at issue567.icry:1:5--1:6
     Unused name: x
+(\x x -> x : [8]) : {a} a -> [8] -> [8]
 [warning] at issue567.icry:2:11--2:12
     This binding for `x` shadows the existing binding at
     issue567.icry:2:5--2:6
-(\x -> \x -> x : [8]) : {a} [8] -> a -> [8]
-[warning] at issue567.icry:3:12--3:13
+[warning] at issue567.icry:2:5--2:6
     Unused name: x
+(\x -> \x -> x : [8]) : {a} a -> [8] -> [8]
 [warning] at issue567.icry:3:12--3:13
     This binding for `x` shadows the existing binding at
     issue567.icry:3:6--3:7
 [warning] at issue567.icry:3:9--3:10
     Unused name: y
-(\(x, y) x -> x) : {a, b, c} (a, b) -> c -> a
+[warning] at issue567.icry:3:6--3:7
+    Unused name: x
+(\(x, y) x -> x) : {a, b, c} (a, b) -> c -> c
 [warning] at issue567.icry:4:13--4:14
     This binding for `x` shadows the existing binding at
     issue567.icry:4:6--4:7
@@ -36,10 +36,10 @@ Loading module Cryptol
     This binding for `x` shadows the existing binding at
     issue567.icry:5:5--5:6
 [warning] at issue567.icry:5:11--5:12
-    Unused name: x
-[warning] at issue567.icry:5:11--5:12
     This binding for `x` shadows the existing binding at
     issue567.icry:5:8--5:9
+[warning] at issue567.icry:5:8--5:9
+    Unused name: x
 [warning] at issue567.icry:5:5--5:6
     Unused name: x
-(\x [x] x -> x) : {a, b, c} a -> [1]b -> c -> b
+(\x [x] x -> x) : {a, b, c} a -> [1]b -> c -> c

